MonoDevelop.MSBuild.Editor.VisualStudio/OpenProjectFile/EditProjectFileCommand.cs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.Shell;
+
+using Task = System.Threading.Tasks.Task;
+
+namespace MonoDevelop.MSBuild.Editor.VisualStudio
+{
+	[Order (10)]
+	[AppliesTo ("OpenProjectFile")]
+	[ExportCommandGroup ("{1496A755-94DE-11D0-8C3F-00C04FC2AAE2}")]
+	internal sealed class EditProjectFileCommand : IAsyncCommandGroupHandler
+	{
+		[Import]
+		private UnconfiguredProject UnconfiguredProject { get; set; }
+
+		[Import (typeof (SVsServiceProvider))]
+		private IServiceProvider ServiceProvider { get; set; }
+
+		[Import]
+		private IProjectThreadingService ProjectThreadingService { get; set; }
+
+		public Task<CommandStatusResult> GetCommandStatusAsync (IImmutableSet<IProjectTree> nodes, long commandId, bool focused, string commandText, CommandStatus progressiveStatus)
+		{
+			if (commandId == 1632) {
+				return Task.FromResult (new CommandStatusResult (true, commandText, CommandStatus.Enabled | CommandStatus.Supported));
+			} else {
+				return Task.FromResult (CommandStatusResult.Unhandled);
+			}
+		}
+
+		public async Task<bool> TryHandleCommandAsync (IImmutableSet<IProjectTree> nodes, long commandId, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut)
+		{
+			if (commandId != 1632) return false;
```

Comment:
Handling the command ("Edit Project File?") isn't the correct approach for this - there are many more ways to open project files than via this command, think File -> Open, double-click from Results view, Error List, Find and Replace - all of which will skip this command and open in the default editor.

Reply:
Double-clicking on the project in the Solution Explorer is handled via a different command. Unfortunately, I don't know how to properly redirect the editor in the other cases you listed. Do you have any suggestions?

Reply:
Scratch that — I had the extension disabled by mistake. File > Open and choosing an MSBuild file opens in the new editor. Preview works properly as well. The only reason we need these commands at all is because the existing commands hard-code the GUID of the built-in XML editor, and this is the only way I know I can override that. AFAIK all other methods of editing an MSBuild file will go through the "modern" editor's way of selecting content types; namely, the static fields in the `MSBuildContentType` class.

Reply:
It looks like we do hardcode the XML editor for both preview, double-click and Edit Project FIle command itself. @davidwengier Given that File - Open doesn't go through that, why do we need to do that? Why can't we go through whatever shell/editor thing is the editor factory?

Reply:
I don't know, but I'm guessing the "why" is because its much easier to just call `OpenDocumentWithSpecific` since you know the editor you want to use (this PR does the same thing for example) than to do all of the right things to find out the editor factory. From memory there is no convenient shell helper method that just finds the best editor factory and opens a file, though I haven't looked for a while and probably last tried before we registered any editor factories at all.

Reply:
We should do exactly what Error List does to open the file.

Reply:
I've fixed this for 16.5. We no longer force the XML editor.

Reply:
Are you sure File -> Open works correctly and opens your editor factory? As far as I can see, we ask the shell to open the XML editor even in that case.